### PR TITLE
Dtls srtp support (RFC 5764)

### PR DESCRIPTION
**Pull request by RonEld**
## Description
This PR supersedes PR #361 by @jeannotlapin, with additional mki support, some rework, and rebased on current development branch

## Status
**IN DEVELOPMENT**

## Requires Backporting
NO  

## Additional comments
Missing still:
tests
interop tests

## Todos
- [ ] Tests
- [ ] Documentation
- [x] Changelog updated
- [ ] Backported

## Steps to test or reproduce
TBD

### CODE REVIEW

File: library/ssl_tls.c [hunk 1]
```diff
@@ -6432,11 +6432,11 @@ mbedtls_ssl_srtp_profile mbedtls_ssl_get_dtls_srtp_protection_profile( const mbe
     return( ssl->dtls_srtp_info.chosen_dtls_srtp_profile);
 }
 
-int mbedtls_ssl_get_dtls_srtp_key_material( const mbedtls_ssl_context *ssl, unsigned char *key, const size_t key_buffer_len, size_t *key_len ) {
+int mbedtls_ssl_get_dtls_srtp_key_material( const mbedtls_ssl_context *ssl, unsigned char *key, size_t *key_len ) {
     *key_len = 0;
```
Comment (RonEld) [inline library/ssl_tls.c]: will be removed, now that key_len is in \ out

File: library/ssl_tls.c [hunk 2]
```diff
@@ -6346,6 +6374,79 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
...
+    return( 0 );
+}
+
+mbedtls_ssl_srtp_profile mbedtls_ssl_get_dtls_srtp_protection_profile( const mbedtls_ssl_context *ssl)
+{
+    return( ssl->dtls_srtp_info.chosen_dtls_srtp_profile);
+}
+
+int mbedtls_ssl_get_dtls_srtp_key_material( const mbedtls_ssl_context *ssl, unsigned char *key, size_t *key_len ) {
+    *key_len = 0;
```
Comment (RonEld) [inline library/ssl_tls.c]: will be removed, now that `key_len` is in \ out

**Comment by RonEld** [comment]
CI failed because of my changes. I will investigate it

**Comment by roxlu** [comment]
@RonEld I was looking into using mbedtls for the SRTP keying material exchange with WebRTC and I was wondering if there are any requirements that I have to make before I can use the features you added? E.g. Do I need to enable the extension?

**Comment by RonEld** [comment]
@roxlu First, this PR is an enhancement to PR #361 by @jeannotlapin , so the credit belongs to them.
Please note that this PR is still no ready for, as CI fails on related issues, and missing tests, and it is prone to changes still.

This PR enables the use_srtp extension in the DTLS handshake , and handles the key material negotiation, but it is not the srtp stack.
Once this PR is ready, and you would want to use srtp-dtls, you will need to enable the extension, and set the srtp profile you would want to use, in preference.

**Comment by roxlu** [comment]
@RonEld yeah I've seen the #361 pull request. I think this pull request has everything I need: exchanging the srtp keying material.  I'm only wondering how to "enable" it and how to extract the keys?  I guess setting the profile is fine. For SRTP I was looking into libsrtp from cisco. 

Any pointers on how to "enable" or use this? Maybe you can past some gist that you used during development?

**Comment by jeannotlapin** [comment]
@roxlu I originally wrote the #361 for linphone/mediastreamer. You can see a real usage of it here:
[dtls_srtp.c](https://github.com/BelledonneCommunications/mediastreamer2/blob/master/src/crypto/dtls_srtp.c).
You'll also need to have a look at this one (it's an abstraction layer mapping to mbedtls) to understand how it works :
[mbedtls.c](https://github.com/BelledonneCommunications/bctoolbox/blob/master/src/crypto/mbedtls.c#L988)

These use the API as I defined it in #361, I did not check if @RonEld modifications changed it.

To enable it you have to make sure the
`#define MBEDTLS_SSL_DTLS_SRTP`
in config.h is not commented.

**Comment by RonEld** [comment]
@roxlu In addition to what I mentioned before, and to what @jeannotlapin meniotned, if you want to use mki value, you should call `mbedtls_ssl_conf_srtp_mki_value_supported()`

**Comment by jeannotlapin** [comment]
@RonEld your commit [55be17f](https://github.com/ARMmbed/mbedtls/pull/1540/commits/55be17f50df5f07b84304109849db16fba4f6af3)  breaks functionality with self signed certificate and thus interoperability with webrtc.
In ssl_srv.c line 2829 you shall also test that ssl->conf->authmode is equal to MBEDTLS_SSL_VERIFY_NONE and in that case only set authmode to MBEDTLS_SSL_VERIFY_OPTIONAL which will force the usage of certificates in the handshake.
Same modification is needed on ssl_tls.c line 4290 and 4438.

Do you want me to do it? I'm not sure I can update this PR.
Also, do you think you will do some more modifications to the API? I'm planning to update the version of mbedtls we use and for that use your PR instead of mine(with the previous fixs) but before that I would like to know how stable the API is.

**Comment by RonEld** [comment]
@jeannotlapin THank you for investigating the failure.
Although this PR got some delays due to other PRs, I did on going back on this feature.
You are welcome to do the fix, if you have the capacity. The reason I modified those parts was because dtls-srtp mandates the certificate verification phase, I haven't looked further though.
Thanks again for you ongoing interest in Mbed TLS!

**Comment by RonEld** [comment]
@jeannotlapin 
>Also, do you think you will do some more modifications to the API?

I don't plan on modifying the API, but since the PR is ongoing and wasn't approved for design yet, it *might* be modified.

**Pull request closed without merging**